include/services/daal_atomic_int.h

```diff
@@ -37,68 +40,293 @@ namespace interface1
+        dataType to_return = my_storage;
+        _ReadWriteBarrier();
+        return to_return;
+    }
+
+    /**
+     * Constructs an atomic object
+     */
+    Atomic() : my_storage(0) {}
+
+    /**
+     * Constructs an atomic object from a value
+     * \param[in] value The value to be assigned to the atomic object
+     */
+    Atomic(dataType value) : my_storage(value) {}
+
+protected:
+    dataType my_storage;
+
+private:
+    Atomic(const Atomic &);
+    Atomic & operator=(const Atomic &);
+};
+#endif // _WIN32 && !_WIN64
+
+#if defined(_WIN64)
 /**
  * <a name="DAAL-CLASS-SERVICES__ATOMIC"></a>
  * \brief Class that represents an atomic object
  *
  * \tparam dataType Data type of the atomic object
  */
-template <typename dataType = int>
+template <typename dataType>
 class DAAL_EXPORT Atomic
+{};
+
+template <>
+class DAAL_EXPORT Atomic<int>
+{
+public:
+    /**
+     * Returns an increment of atomic object
+     * \return An increment of atomic object
+     */
+    inline int inc()
+    {
+        DAAL_ASSERT(sizeof(my_storage) == sizeof(long))
+        return (int)(_InterlockedExchangeAdd((long *)(&my_storage), 1) + 1);
+    }
+
+    /**
+     * Returns a decrement of atomic object
+     * \return An decrement of atomic object
+     */
+    inline int dec()
+    {
+        DAAL_ASSERT(sizeof(my_storage) == sizeof(long))
+        return (int)(_InterlockedExchangeAdd((long *)(&my_storage), -1) - 1);
+    }
+
+    /**
+     * Assigns the value to atomic object
+     * \param[in] value    The value to be assigned
+     */
+    inline void set(int value)
+    {
+        _ReadWriteBarrier();
+        my_storage = value;
+    }
+
+    /**
+     * Returns the value of the atomic object
+     * \return The value of the atomic object
+     */
+    inline int get() const
+    {
+        int to_return = my_storage;
+        _ReadWriteBarrier();
+        return to_return;
+    }
+
+    /**
+     * Constructs an atomic object
+     */
+    Atomic() : my_storage(0) {}
+
+    /**
+     * Constructs an atomic object from a value
+     * \param[in] value The value to be assigned to the atomic object
+     */
+    Atomic(int value) : my_storage(value) {}
+
+protected:
+    int my_storage;
+
+private:
+    Atomic(const Atomic &);
+    Atomic & operator=(const Atomic &);
+};
+
+template <>
+class DAAL_EXPORT Atomic<size_t>
 {
 public:
     /**
      * Returns an increment of atomic object
      * \return An increment of atomic object
      */
-    dataType inc();
+    inline size_t inc()
+    {
+        DAAL_ASSERT(sizeof(my_storage) == sizeof(size_t))
+        return (size_t)(_InterlockedExchangeAdd64((__int64 *)(&my_storage), 1) + 1);
+    }
 
     /**
      * Returns a decrement of atomic object
      * \return An decrement of atomic object
      */
-    dataType dec();
+    inline size_t dec()
+    {
+        DAAL_ASSERT(sizeof(my_storage) == sizeof(size_t))
+        return (size_t)(_InterlockedExchangeAdd64((__int64 *)(&my_storage), -1) - 1);
```

Comment:
The result can be wrong if highest bit is 1 in the original number. Also, it violates SDL.

Reply:
I don`t see problems here:
1. we don`t loose number sign here - no sdl violations
2. 0x80000001. size_t(32bit) is 2147483649. int(32bit) is -2147483647. int(-2147483647 - 1) = 0x80000001 = size_t(2147483649 - 1)